#! /bin/sh /usr/share/dpatch/dpatch-run
## 99_security_status_cgi_servicegroup.dpatch by Jonas Meurer <jmeurer@inet.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Stop cgi-bin/status.c from listing unauthorized hosts and
## DP:				services in servicegroup view
## DP: Upstream bugreport: http://tracker.nagios.org/view.php?id=456

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' nagios3-3.4.1~/cgi/status.c nagios3-3.4.1/cgi/status.c
--- nagios3-3.4.1~/cgi/status.c	2012-02-13 21:40:42.000000000 +0100
+++ nagios3-3.4.1/cgi/status.c	2013-06-26 16:52:37.668132234 +0200
@@ -2534,6 +2534,10 @@
 		if(temp_host == NULL)
 			continue;
 
+		/* make sure user has rights to view this host */
+		if(is_authorized_for_host(temp_host, &current_authdata) == FALSE)
+			continue;
+
 		/* skip this if it isn't a new host... */
 		if(temp_host == last_host)
 			continue;
@@ -2739,6 +2743,10 @@
 		if(temp_host == NULL)
 			continue;
 
+		/* make sure user has rights to view this host */
+		if(is_authorized_for_host(temp_host, &current_authdata) == FALSE)
+			continue;
+
 		/* skip this if it isn't a new host... */
 		if(temp_host == last_host)
 			continue;
@@ -2918,6 +2926,10 @@
 		if(temp_service == NULL)
 			continue;
 
+		/* make sure user has rights to view this service */
+		if(is_authorized_for_service(temp_service, &current_authdata) == FALSE)
+			continue;
+
 		/* find the service status */
 		temp_servicestatus = find_servicestatus(temp_service->host_name, temp_service->description);
 		if(temp_servicestatus == NULL)
@@ -3270,6 +3282,10 @@
 		if(temp_host == NULL)
 			continue;
 
+		/* make sure user has rights to view this host */
+		if(is_authorized_for_host(temp_host, &current_authdata) == FALSE)
+			continue;
+
 		/* get the status of the host */
 		temp_hoststatus = find_hoststatus(temp_host->name);
 		if(temp_hoststatus == NULL)
